Add a hidden field on the collection_radio_buttons

maurogeorge opened the pull request:
The same behavior that occurs on file_field(https://github.com/rails/rails/issues/17947) occurs with the `collection_radio_buttons`.

maurogeorge committed:
Add a hidden field on the collection_radio_buttons

This will avoid a error be raised when the only input on the form is the
`collection_radio_buttons`.

rafaelfranca commented on actionview/lib/action_view/helpers/tags/collection_radio_buttons.rb:
```diff
@@ -14,7 +14,8 @@ def radio_button(extra_html_options={})
 
         def render(&block)
-          render_collection do |item, value, text, default_html_options|
+          options = @options.stringify_keys
```
I think we can simplify the implementation of this method and collection_check_boxes moving it `render_collection`. Also I don't think we need to stringify in this case.

rafaelfranca commented on actionview/lib/action_view/helpers/tags/collection_radio_buttons.rb:
```diff
@@ -30,6 +39,18 @@ def render(&block)
             builder.radio_button + builder.label
           end
+
+          def hidden_field
```
This should go to a shared place too.

rafaelfranca commented on actionview/CHANGELOG.md:
```diff
@@ -12,5 +12,9 @@
 
     *Angelo Capilleri*
 
+*   Add a `hidden_field` on the `collection_radio_buttons` to avoid raise a error
```
It should be at top of the file

rafaelfranca commented on actionview/test/template/form_collections_helper_test.rb:
```diff
@@ -198,6 +198,41 @@ def with_collection_check_boxes(*args, &block)
+    collection = [Category.new(1, 'Category 1'), Category.new(2, 'Category 2')]
+    with_collection_radio_buttons :user, :category_ids, collection, :id, :name
+
+    assert_select "input[type=hidden][name='user[category_ids][]'][value='']", :count => 1
```
Use new hash syntax

rafaelfranca commented on actionview/test/template/form_collections_helper_test.rb:
```diff
@@ -198,6 +198,41 @@ def with_collection_check_boxes(*args, &block)
+
+  test 'collection radio buttons generates a hidden field using the given :name in :html_options' do
+    collection = [Category.new(1, 'Category 1'), Category.new(2, 'Category 2')]
+    with_collection_radio_buttons :user, :category_ids, collection, :id, :name, {}, {name: "user[other_category_ids][]"}
```
space after the `{` and before `}` at the last hash.

rafaelfranca commented on actionview/test/template/form_collections_helper_test.rb:
```diff
@@ -198,6 +198,41 @@ def with_collection_check_boxes(*args, &block)
+    collection = [Category.new(1, 'Category 1'), Category.new(2, 'Category 2')]
+    with_collection_radio_buttons :user, :category_ids, collection, :id, :name, {}, {name: "user[other_category_ids][]"}
+
+    assert_select "input[type=hidden][name='user[other_category_ids][]'][value='']", :count => 1
```
New hash syntax

rafaelfranca commented on actionview/test/template/form_collections_helper_test.rb:
```diff
@@ -198,6 +198,41 @@ def with_collection_check_boxes(*args, &block)
+
+  test 'collection radio buttons does not generate a hidden field if include_hidden option is false with key as string' do
+    collection = [Category.new(1, 'Category 1'), Category.new(2, 'Category 2')]
+    with_collection_radio_buttons :user, :category_ids, collection, :id, :name, 'include_hidden' => false
```
Not sure if this is needed. Does collection_check_box have this test?

eileencodes commented on actionview/lib/action_view/helpers/form_options_helper.rb:
```diff
@@ -644,6 +644,24 @@ def time_zone_options_for_select(selected = nil, priority_zones = nil, model = :
+      # The HTML specification says when nothing is select on a collection of radio buttons
+      # web browsers do not send any value to server.
+      # Unfortunately this introduces a gotcha:
+      # if an +User+ model has a +category_ids+ field, and in the form none category is selected no +category_ids+ parameter is sent. So,
```
This should be "If a +User+" because the way "user" is pronounced is an exception to the article "an" before a vowel rule because it makes the same sound as "y" in "you".

maurogeorge replied:
Thanks, I will update.

maurogeorge committed:
Regression test on CollectionCheckBoxes to accept  `include_hidden`
as string

maurogeorge commented:
@rafaelfranca I updated the code following your suggestions.

About `collection_check_box` it dont have the test of `include_hidden` as string, I added the missing test, now both works with the string. I added the stringfy to fix this.

maurogeorge commented:
@rafaelfranca when you have some time could you please review :smile:

maurogeorge commented:
@rafaelfranca fixed the conflicts, when you have some time could you please review

maurogeorge commented:
@rafaelfranca please review when you have some time :smile:

maurogeorge commented:
@rafaelfranca I rebased again to fix the conflicts, please review when you have some time.Is there anything I can do to help? @carlosantoniodasilva maybe you can review this :smile:

claudiob commented on actionview/lib/action_view/helpers/tags/collection_helpers.rb:
```diff
@@ -79,6 +79,39 @@ def render_collection #:nodoc:
+          end
+        end
+
+        def hidden_field #:nodoc:
```
Hello @maurogeorge – I'm :+1: with this PR :grinning: 

Since you added `def hidden_field`, could the one [already defined in collection_check_boxes.rb](https://github.com/rails/rails/pull/18303/files#diff-ad5a6748d86270f717368d13026b475bR27) be removed?

maurogeorge replied:
Hi @claudiob great to hear this :smile: 

You are right I can remove the mentioned `hidden_field`. I updated the code and rebased.

rafaelfranca commented on actionview/lib/action_view/helpers/tags/collection_check_boxes.rb:
```diff
@@ -9,29 +9,13 @@ class CollectionCheckBoxes < Base # :nodoc:
 
         def render(&block)
-          rendered_collection = render_collection do |item, value, text, default_html_options|
-            default_html_options[:multiple] = true
```
Multiple is not being handled anymore with your refactoring.

maurogeorge replied:
@rafaelfranca I moved the multiple options to the [`check_box`](https://github.com/rails/rails/pull/18303/files#diff-ad5a6748d86270f717368d13026b475bR12) it is covered by the tests, if I remove this line I got some broken tests

rafaelfranca commented on actionview/lib/action_view/helpers/tags/collection_helpers.rb:
```diff
@@ -79,6 +79,39 @@ def render_collection #:nodoc:
+        def hidden_field #:nodoc:
+          hidden_name = @html_options[:name]
+
+          hidden_name ||= if @options.has_key?(:index)
```
Why this is not using `tag_name`?

maurogeorge replied:
Thanks for the tip, I updated the code to use `tag_name`.

rafaelfranca replied:
Does not the previous implementation work? Does it need to check `:index` option?

claudiob commented on actionview/CHANGELOG.md:
```diff
@@ -1,8 +1,13 @@
-     
-     Example: 
-        
+*   Add a `hidden_field` on the `collection_radio_buttons` to avoid raise a error
```
There is something strange in this change… it should only show your new lines, not replace the existing one.
Did you accidentally edit the next CHANGELOG message too?

If you review this, also fix `avoid raise` to `avoid raising` :wink:

claudiob commented on actionview/lib/action_view/helpers/form_options_helper.rb:
```diff
@@ -644,6 +644,24 @@ def time_zone_options_for_select(selected = nil, priority_zones = nil, model = :
+      # every collection of radio buttons. The hidden field has the same name as collection radio button and blank value.
+      #
+      # In case if you don't want the helper to generate this hidden field you can specify
+      # <tt>include_hidden: false</tt> option.
```
Hello @maurogeorge – I think this comment is incorrect.

In the case of a collection of radio button, you don't have `category_ids` (plural) and mass assignments.

Instead, the effect of this commit is to tackle the case where you have a form that contains *only* a collection of radio buttons (and no other field type).

For instance, if you have a User#update form, where all you can do is select the category with radio buttons.

If you submit such a form without selecting any radio button, then you expect the params to contain `{user: {category_id: nil}`. This lets you save the object or optionally raise a validation error for the absence of category.

However, the current behavior is that the params will not contain `{user: ...}` at all. 
Therefore if your controller requires strong parameters like `params.require(:user).permit(...)`, the `require` part will fail and the user will see a 400 error.

maurogeorge commented:
@claudiob @rafaelfranca thanks a lot for the review :heart: :heart_eyes: 
Sorry for the delay, I was involved with other things :cry: 
I updated the code, please take a look when you have some time :smile:

maurogeorge commented:
Thanks for the merge :heart:
